test/src/test/java/org/corfudb/integration/WorkflowIT.java

```diff
@@ -304,4 +304,86 @@ public void forceRemoveTwoDeadNodes() throws Exception {
+     * Now 2 new nodes node_2 and node_3 are added to the cluster after which node_1 is shutdown.
+     * Finally we should see that the trimMark should be updated on the new nodes and the
+     * FastObjectLoader trying to recreate the state from these 2 nodes should be able to do so.
+     */
+    @Test
+    public void addNodeWithTrim() throws Exception {
+        Harness harness = new Harness();
+
+        final int numNodes = 3;
+        final int PORT_1 = 9001;
+        final int PORT_2 = 9002;
+        List<Node> nodeList = harness.deployCluster(1);
+        Node n0 = nodeList.get(0);
+        Node n1 = harness.deployUnbootstrappedNode(PORT_1);
+        Node n2 = harness.deployUnbootstrappedNode(PORT_2);
+
+        CorfuRuntime rt = new CorfuRuntime(n0.getClusterAddress()).connect();
+        CorfuTable<String, Integer> table = rt.getObjectsView()
+                .build()
+                .setType(CorfuTable.class)
+                .setStreamName("test")
+                .open();
+        final int entriesCount = 1_000;
+
+        // Write 1_000 entries.
+        for (int i = 0; i < entriesCount; i++) {
+            table.put(Integer.toString(i), i);
+        }
+
+        // Checkpoint and trim the entries.
+        MultiCheckpointWriter mcw = new MultiCheckpointWriter();
+        mcw.addMap(table);
+        long prefixTrimAddress = mcw.appendCheckpoints(rt, "author");
+        rt.getAddressSpaceView().prefixTrim(prefixTrimAddress);
+        rt.getAddressSpaceView().invalidateServerCaches();
+        rt.getAddressSpaceView().invalidateClientCache();
+        rt.getAddressSpaceView().gc();
+
+        // Write another batch of 1_000 entries.
+        for (int i = 0; i < entriesCount; i++) {
+            table.put(Integer.toString(i), i);
+        }
+
+        // Add 2 new nodes.
+        final int retries = 3;
+        final Duration timeout = PARAMETERS.TIMEOUT_LONG;
+        final Duration pollPeriod = PARAMETERS.TIMEOUT_VERY_SHORT;
+        rt.getManagementView().addNode("localhost:9001", retries, timeout, pollPeriod);
+        rt.getManagementView().addNode("localhost:9002", retries, timeout, pollPeriod);
+
+        rt.invalidateLayout();
+        Layout layoutAfterAdds = rt.getLayoutView().getLayout();
+        assertThat(layoutAfterAdds.getSegments().stream()
+                .allMatch(s -> s.getAllLogServers().size() == numNodes)).isTrue();
+
+        run(n0.shutdown);
+
+        for (int i = 0; i < PARAMETERS.NUM_ITERATIONS_MODERATE; i++) {
+            if (rt.getLayoutView().getLayout().getEpoch() > layoutAfterAdds.getEpoch()) {
+                break;
+            }
+            rt.invalidateLayout();
+            Sleep.sleepUninterruptibly(PARAMETERS.TIMEOUT_SHORT);
+        }
+
+        // Assert that the new nodes should have the correct trimMark.
+        assertThat(rt.getLayoutView().getRuntimeLayout().getLogUnitClient("localhost:9001").getTrimMark().get())
```

Comment:
Also test the address space get trim mark method.

Reply:
Done.